drivers: video: video_mcux_smartdma: add SMARTDMA video driver

danieldegrasse opened the pull request:
Add SMARTDMA video driver. This driver uses the SMARTDMA engine as a
parallel camera interface, which can read QVGA frames from a camera
device. Due to SRAM constraints, the video driver divides the camera
stream into multiple horizontal video buffers as it streams them back to
an application.

Note that this PR also contains two additional major changes:
- an addition to the video API to enable streaming partial frames within one video buffer
- a rework of the SMARTDMA driver interface to better support custom firmware use cases like the camera in this PR

~This PR depends on #72826 for camera support~

PR has been tested on the following boards:
- `frdm_mcxn947//cpu0`: `samples/subsys/video/capture`
- `mimxrt595_evk//cm33`: `samples/subsys/display/lvgl` (using G1120B0MIPI shield)

danieldegrasse committed:
drivers: dma: dma_mcux_smartdma: update interface to support custom FW

The SMARTDMA is a programmable DMA engine, and supports custom firmware
in order to run complex DMA operations. Update the driver to increase
the flexibility users have when configuring the SMARTDMA with
custom firmware, and remove the RT500 display firmware specific
definitions and functionality from the driver.

This display setup is now handled from the MIPI DSI driver, since the
firmware used for this case is specific to the MIPI DSI IP.

This change also requires an update to the RT500 devicetree, as the
register definition for the SMARTDMA has changed, so the base address
must as well.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
drivers: video: video_mcux_smartdma: add SMARTDMA video driver

Add SMARTDMA video driver. This driver uses the SMARTDMA engine as a
parallel camera interface, which can read QVGA frames from a camera
device. Due to SRAM constraints, the video driver divides the camera
stream into multiple horizontal video buffers as it streams them back to
an application.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
dts: arm: nxp: nxp_mcxn94x: add definition for SMARTDMA device

Add definition for SMARTDMA device to the MCXN94x devicetree.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
boards: nxp: frdm_mcn947: add support for LPI2C7

Add support for LPI2C7. This peripheral is used for interfacing with
cameras connected to the J9 header.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
boards: nxp: frdm_mcnx947: add support for SMARTDMA camera engine

Add support for SMARTDMA camera engine, and a OV7670 parallel camera
definition for the frdm_mcnx947 board. This support has been tested with
the video capture sample.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
samples: video: capture: add support for SMARTDMA

Add support for using the SMARTDMA engine on the FRDM-MCXN947 board with
the video capture sample.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

josuah commented:
> * an addition to the video API to enable streaming partial frames within one video buffer

Some news about it there, with a proposed API: https://github.com/zephyrproject-rtos/zephyr/pull/66994#issuecomment-1893401899

josuah commented:
[here](https://github.com/zephyrproject-rtos/zephyr/pull/66994/files#diff-a6394ee84ba37d741b1aecea29085811d0f447fe128a655c3840eb59c3e86604R826-R831) (need to click on "load diff") is a new API that seems underway:
```c
	if (drv_data->fifo_length == 0) {
		vbuf->flags = VIDEO_BUF_EOF;
	} else {
		vbuf->flags = VIDEO_BUF_FRAG;
		k_work_submit_to_queue(&ac_work_q, &drv_data->buf_work);
	}
```

danieldegrasse commented:
> [here](https://github.com/zephyrproject-rtos/zephyr/pull/66994/files#diff-a6394ee84ba37d741b1aecea29085811d0f447fe128a655c3840eb59c3e86604R826-R831) (need to click on "load diff") is a new API that seems underway:

No issues with this API, I suppose we would like to make this PR dependent on that API merging? I could also cherry pick the commits adding the API from that PR (@Arducam would you be ok with that approach?)

ArduCAM commented:
> > [here](https://github.com/zephyrproject-rtos/zephyr/pull/66994/files#diff-a6394ee84ba37d741b1aecea29085811d0f447fe128a655c3840eb59c3e86604R826-R831) (need to click on "load diff") is a new API that seems underway:
> 
> No issues with this API, I suppose we would like to make this PR dependent on that API merging? I could also cherry pick the commits adding the API from that PR (@ArduCAM would you be ok with that approach?)

Yes, have marked resolved

danieldegrasse commented:
> @danieldegrasse , the HAL side PR merged a while ago. I suspect you don't even need to include a manifest update anymore.

Updated to include SHA instead of PR reference

danieldegrasse commented:
@loicpoulain can you take a look at this? Would like your thoughts on the API changes here

josuah commented:
I wonder if it could be summarized to a `min_vbuf_size` and `max_vbuf_size`, covering most/all buffering restrictions:

Devices that **produce** data (OUT endpoints), `max_vbuf_size` is a hint for allocating memory without wasting.

- If it supports **only full frames**, `min_vbuf_size` is the frame size.
  -> The application receives buffer with EOF flag.
- If it supports **fragmented** frames with a **arbitrary** buffer size, `min_vbuf_size=1`.
  -> The application receives buffers with FRAG only if it is too small, or only EOF when it always fit.
- If it supports **fragmented** frames with a **minimum** buffer size, `min_vbuf_size=that`.
  -> The application receives buffers with `bytesused == size`.
- If it supports **fragmented** frames with an **exact** buffer size, `min_vbuf_size=that`.
  -> The application receives buffers with `bytesused <= size`.

Devices that **consume** data (IN endpoints):

- If it supports **only full frames**, `min_vbuf_size` = `max_vbuf_size` = frame size.
  -> The application needs to accumulate FRAG frames until it has `min_vbuf_size` before sending it.
- If it supports **fragmented** frames with a **arbitrary** buffer size, `min_vbuf_size=1`, `min_vbuf_size=frame_size`.
  -> The application needs to accumulate FRAG frames until it has `min_vbuf_size` before sending it.
- If it supports **fragmented** frames with a **minimum** buffer size, `min_vbuf_size=that`, `max_vbuf_size=frame_size`.
  -> The application needs to accumulate FRAG frames until it has `min_vbuf_size` before sending it.
- If it supports **fragmented** frames with a **maximum** buffer size, `max_vbuf_size=that`, `min_vbuf_size=1`.
  -> The application will need to cut the buffer to `max_vbuf_size`, remainder put on next frame.
- If it supports **fragmented** frames with an **exact** buffer size, `min_vbuf_size=max_vbuf_size`.
  -> The application needs to accumulate FRAG frames until it has `min_vbuf_size` before sending it.
  -> The application will need to cut the buffer to `max_vbuf_size`, remainder put on next frame.

Applications can support every use-case without introducing flags:

- Accumulate FRAG frames until it has `min_vbuf_size`.
- Cut the buffer to `max_vbuf_size`, remainder put on next frame.

danieldegrasse commented:
> If it supports fragmented frames with a minimum buffer size, min_vbuf_size=that.
-> The application receives buffers with bytesused == size.
If it supports fragmented frames with an exact buffer size, min_vbuf_size=that.
-> The application receives buffers with bytesused <= size.

I'm not sure I understand why these cases are setup the way they are. Wouldn't the application use "min_vbuf_size" to determine the size of buffer it should allocate for the video input device to fill? In that case, wouldn't the cases work like so?

- If it supports fragmented frames with a minimum buffer size, min_vbuf_size=that.
-> The application receives buffers with **bytesused <= size**.
-If it supports fragmented frames with an exact buffer size, min_vbuf_size=that.
-> The application receives buffers with **bytesused == size**.

Overall, I like this idea. The key thing we need to capture is a way for applications to easily determine the optimal buffer size to provide to the video device, which this does do. An application looking to minimize allocations for a video input device could allocate buffers equal to "max_vbuf_size", right?

josuah commented:
Sorry for the long wait, and thank you for your answer!

> Wouldn't the application use "min_vbuf_size" to determine the size of buffer it should allocate for the video input device to fill?

There might be extreme cases like `min_vbuf_size=1` but yes.

> `min_vbuf_size=minimum buffer size` => `bytesused <= size`

Good point! I missed this.

> `min_vbuf_size=exact buffer size` => `bytesused == size`

Maybe the last fragment would be exceptinoally shorter than `min_vbuf_size` for some hardware?
For instance JPEG will have non-exact frame sizes.

> An application looking to minimize allocations for a video input device could allocate buffers equal to "max_vbuf_size", right?

That is how I was picturing it as well.

danieldegrasse commented:
> Sorry for the long wait, and thank you for your answer!
> 
> > Wouldn't the application use "min_vbuf_size" to determine the size of buffer it should allocate for the video input device to fill?
> 
> There might be extreme cases like `min_vbuf_size=1` but yes.
> 
> > `min_vbuf_size=minimum buffer size` => `bytesused <= size`
> 
> Good point! I missed this.
> 
> > `min_vbuf_size=exact buffer size` => `bytesused == size`
> 
> Maybe the last fragment would be exceptinoally shorter than `min_vbuf_size` for some hardware? For instance JPEG will have non-exact frame sizes.
> 
> > An application looking to minimize allocations for a video input device could allocate buffers equal to "max_vbuf_size", right?
> 
> That is how I was picturing it as well.

Okay, then this all makes sense. Another thing we need to consider (which this PR doesn't currently capture is that for video endpoints producing partial framebuffers, it would be useful for the endpoint to be able to indicate *where* in the frame's progression a buffer exists. IE for a device that streams fixed size frames 1/8 the size of the full camera frame, indicating which of the 8 positions that frame exists in. This should probably be encoded as a pixel offset for maximum flexibility though.

The use case I see for this is to handle cases where the consumer can't keep up with the video producer, and some frame fragments are dropped. Say that a camera output is being streamed to a display, knowing *where* on the display to place the next frame fragment will be useful if the application ends up dropping a frame from the video producer.

Perhaps we make the following changes?
```
/* Extend video capabilities with minimum/maximum video buffer size */
struct video_caps {
	/** list of video format capabilities (zero terminated). */
	const struct video_format_cap *format_caps;
	/** minimal count of video buffers to enqueue before being able to start
	 * the stream.
	 */
	uint8_t min_vbuf_count;
	/** Minimum size of video buffer the application may provide to the video
	 * endpoint
	 */
	min_vbuf_size;
	/** Maximum size of video buffer the application may provide to the video
	 * endpoint
	 */
	max_vbuf_size;
};

/* Extend video buffer structure with frame-offset field */
struct video_buffer {
	/** pointer to driver specific data. */
	void *driver_data;
	/** pointer to the start of the buffer. */
	uint8_t *buffer;
	/** size of the buffer in bytes. */
	uint32_t size;
	/** number of bytes occupied by the valid data in the buffer. */
	uint32_t bytesused;
	/** time reference in milliseconds at which the last data byte was
	 * actually received for input endpoints or to be consumed for output
	 * endpoints.
	 */
	uint32_t timestamp;
	/** Y offset in pixels that this buffer fragment starts at within the frame,
	 * from the beginning of the frame
	 */
	uint32_t frame-offset;
};
```

josuah commented:
Sorry for the delay again!

> The use case I see for this is to handle cases where the consumer can't keep up with the video producer, and some frame fragments are dropped.

This sounds a very likely thing to happen in real-world systems!

> knowing where on the display to place the next frame fragment will be useful if the application ends up dropping a frame from the video producer.

Does dropping the whole frame seem like a reasonable compromise? It would then suffice to wait the next EOF flag and start from 0. Or was it about recovering a feed mid-frame to overcome some more severe cases?

josuah commented:
Ah my bad... without the frame-offset, the sink would never know if two fragment are in the right order or if some was missed. As all it sees would be a big block of pixel.

An alternative of `timestamp` would be `frame_number`: easier to count lost frames, not straining the time API, but not helping with end-to-end latency or synchronizing a display of a feed through time in case it is doing bursts of data then pauses.

danieldegrasse commented:
> Ah my bad... without the frame-offset, the sink would never know if two fragment are in the right order or if some was missed. As all it sees would be a big block of pixel.
> 
>An alternative of timestamp would be frame_number: easier to count lost frames, not straining the time API, but not helping with end-to-end latency or synchronizing a display of a feed through time in case it is doing bursts of data then pauses.

Would `frame_number` work here though? The issue I see is that there is a distinction between the *fragment number* of the frame and the *frame number* of the frame. Some sinks probably need to be able to count both.

DerekSnell commented on samples/subsys/video/capture/boards/frdm_mcxn947_cpu0.overlay:
Hi @danieldegrasse ,
This sample was moved in the tree, and is now at [samples/drivers/video/capture](https://github.com/zephyrproject-rtos/zephyr/tree/main/samples/drivers/video/capture).  So these 3 files need to be moved.

danieldegrasse replied:
Thanks for the heads up- awaiting feedback from @josuah on the proposed API changes here: https://github.com/zephyrproject-rtos/zephyr/pull/72827#issuecomment-2302144202, once we settle on an approach there I can update the PR

josuah replied:
It all makes sense, thank you for explaining all the details, this will help while applying this new API for future or existing drivers.  Pursuing on the main thread,

josuah commented on drivers/video/video_mcux_csi.c:
```diff
@@ -293,7 +293,8 @@ static int video_mcux_csi_enqueue(const struct device *dev, enum video_endpoint_
 	to_read = data->csi_config.linePitch_Bytes * data->csi_config.height;
-	vbuf->bytesused = to_read;
+	vbuf->bytesframe = vbuf->bytesused = to_read;
+	vbuf->flags = VIDEO_BUF_EOF;
```
Maybe other pull requests would be needed to propagate this change to the other drivers too?
So that it gets possible to rely on this from application/subsystem code.

josuah reviewed (dismissed):
Marking "approve" for the ongoing changes, though letting you pursue with the extra APIs as well.

Should the extra APIs be introduced on a separate PR? This could other vendors to integrate the changes.

mmahadevan108 commented:
@danieldegrasse , can you help rebase

dleach02 commented:
> Marking "approve" for the ongoing changes, though letting you pursue with the extra APIs as well.
> 
> Should the extra APIs be introduced on a separate PR? This could other vendors to integrate the changes.

@danieldegrasse any comment on @josuah question?

josuah commented:
Side note: the fragmentation APIs introduced do make sense to me, and I even started to experiment with them in #63421

danieldegrasse commented:
> Marking "approve" for the ongoing changes, though letting you pursue with the extra APIs as well.
> 
> Should the extra APIs be introduced on a separate PR? This could other vendors to integrate the changes.

We need to introduce some level of API changes in this PR- the camera engine here fundamentally *can't* stream a full frame into memory on the MCXN947- the SRAM is too small to support it.

Are you ok with the changes proposed in this comment: https://github.com/zephyrproject-rtos/zephyr/pull/72827#issuecomment-2302144202? If so, I can update the PR and get things moving forwards

josuah commented:
> Are you ok with the changes proposed in this comment: [#72827 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/72827#issuecomment-2302144202)? If so, I can update the PR and get things moving forwards

Yes absolutely, this looks like working! Sorry if my position was ambiguous.
For uncompressed formats I suppose `pitch * frame_offset` gives the position in bytes.
Thank you for the time spent organizing this!

danieldegrasse commented:
> Yes absolutely, this looks like working! Sorry if my position was ambiguous.
For uncompressed formats I suppose pitch * frame_offset gives the position in bytes.
Thank you for the time spent organizing this!

No worries, I'll work on getting this PR updated

danieldegrasse committed:
drivers: video: add min_line_count and max_line_count to video_caps

Add min and max line count to the video_caps structure- these fields can
be used by applications to determine the size of video buffer they need
to allocate for a video endpoint

The min and max line count fields are designed to enable supporting
endpoints that may produce or consume partial frames within each video
buffer, and may support arbitrarily sized video buffers.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
drivers: video: update video drivers to handle min/max line count

Update existing video drivers to handle the min/max line count field
within the video_caps structure. All drivers work with full frames
currently, so use the special LINE_COUNT_HEIGHT value to indicate this.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
samples: drivers: video: update video samples for min/max line count

Update video samples to use min/max line count capabilities when
allocating video buffers.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
drivers: video: add line_offset field to video_buffer structure

Add line_offset field to the video_buffer structure. This field
indicates the offset (in horizontal lines) within a frame that a video
buffer starts at. This is useful for video devices that produce or
consume video buffers that constitute a partial frame.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
drivers: video: handle line_offset field

Since all video drivers in tree use a full frame, their video buffers
will always start at a line_offset of 0 within the frame.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
samples: drivers: video: handle line_offset field

Handle line_offset field within video capture sample. Other samples do
not support partial framebuffers and therefore can ignore this field.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
samples: drivers: video: capture: don't apply format unless needed

Do not apply format setting unless needed. Also, correct the check for
the RGB565 format setting- the zephyr display API treats RGB565 and
BGR565 as big endian, so the format needed here is BGR565.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse commented:
> No worries, I'll work on getting this PR updated

PR has been updated- something I realized while working on this was that we can't really provide framebuffer size bounds until the format is set- therefore, I chose to report the min/max number of *frame lines* rather than the buffer size, as this should be independent of format (I think).

danieldegrasse committed:
drivers: video: ov7670: improve resolution init code

Improve resolution init code for OV7670 driver, to properly program the
full set of registers needed to realize a given output resolution. The
settings for these registers are based on those used in the MCUX SDK
driver, which are derived from the resolution register settings given in
the OV7670 programming guide, with a different configuration for the
CLKSRC and DBLV registers used for the input clock.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

loicpoulain reviewed (dismissed):
Looks good, is it ok to merge @danieldegrasse ?

josuah reviewed (dismissed):
It is true that many video processing cores have a line buffer, or a 5-line buffers (i.e. small [kernels](https://en.wikipedia.org/wiki/Kernel_(image_processing))) rather than an XYZ bytes buffers.

A line-based min/max size sounds good.

josuah reviewed (approved):
Thank you for adjusting all the other drivers too!
